### PR TITLE
Replace bp/sp-vars by generic stack-based ones

### PR DESCRIPTION
**Pull request by thestr4ng3r**
Pending updates in other projects to be merged after this:
https://github.com/rizinorg/rz-ghidra/pull/307
https://github.com/rizinorg/rz-retdec/pull/16
https://github.com/rizinorg/cutter/pull/3062
https://github.com/rizinorg/jsdec/pull/26

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
Variables on the stack are not identified by bp/sp+offset anymore, but
by their address from the bottom of the stack frame (RzStackAddr),
independent of how they are accessed.
So now there are only two kinds of variables: stack and register.

This required some major refactoring and other changes:
* RzAnalysisVar.isarg was removed. Whether a variable is an argument is
  now specified implicitly by its storage location.
* Varsub of struct fields had to be rewritten so fields can be queried
  by arbitrary stack addresses using the recently introduced sp
  tracking, as the old approach to fill a list with all fields would not
  work anymore.
* analysis.vars.stackname was removed, new behavior is more similar to
  this being true before.
* Variables will not be created at stack+0 now, because the return
  address is there. Before, vars were only created sometimes in such
  cases.
* Variables created from bp offsets in x86 are not deleted anymore if
  the function's bp_frame is false (see removed
  rz_analysis_function_delete_vars_by_kind(fcn,
  RZ_ANALYSIS_VAR_KIND_BPV); calls). This may lead to some
  false-positive detected variables. Whether this really is a practical
  issue is yet to be seen. At least there are no meaningful tests that
  are broken by this.
* Applying variables from dwarf needed some fixes for determining the
  correct stack locations of variables in order to write meaningful
  tests. The handling is still not entirely correct for all
  possibilities of dwarf info, but at least the changed/added test cases
  are right and serve as a reference for future changes.
* Projects version 11 is introduced.
* afvb commands have been removed, afvs now handles all stack vars.

The key change is the diff in `RzAnalysisVar` (see in particular `RzAnalysisVarStorage`).

Before:
Bildschirm­foto 2022-12-25 um 20 35 42

After:
Bildschirm­foto 2022-12-25 um 20 36 24

### CODE REVIEW

File: changenotes.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+analysis.vars.stackname removed, new behavior is more similar to this being true
```
Comment (ret2libc) [inline changenotes.txt]: Is this a temporary file?

File: librz/include/rz_analysis.h [hunk 2]
```diff
@@ -614,19 +634,39 @@ typedef enum {
...
+	 * Register used for access.
+	 * For example when accessing some stack variable by `[rbp - 0x10]`, this will be "rbp"
+	 */
+	const char *reg;
+
+	/**
+	 * Delta added to register when the var is accessed,
+	 * For example for `[rbp - 0x10]`, this will be -0x10.
+	 */
+	st64 addend;
```
Comment (wargio) [inline librz/include/rz_analysis.h]: maybe `relative_offset` is a better name ?

Comment (thestr4ng3r) [reply librz/include/rz_analysis.h]: I chose this name to deliberately express that is is a value that is added to the register, in direct contrast to the stack offset that is saved in the variable itself and describes the offset from the bottom of the stack.

Comment (wargio) [reply librz/include/rz_analysis.h]: i still don't like it, but makes sense.

File: librz/include/rz_analysis.h [hunk 3]
```diff
@@ -600,11 +599,30 @@ typedef struct rz_analysis_bind_t {
...
-	RZ_ANALYSIS_VAR_KIND_SPV = 's'
-} RzAnalysisVarKind;
+/**
+ * \brief An address on the stack
+ *
+ * These addresses are a relative offset to the value of the stack pointer
+ * when entering a function, i.e. just before the first instruction inside the function is executed.
+ * Thus, with the most common calling conventions, the return address with be at addr 0,
+ * arguments at positive addresses and local variables at negative addresses:
+ *
```
Comment (ret2libc) [inline librz/include/rz_analysis.h]: So it means that if I have something like:
```
push r15
push r14
push rbx
sub rsp, 0x50
mov qword [rsp + 10], 0x22
```

the variable at `rsp+10` will be marked as *-88*, right? 8(r15) + 8 (r14) + 8(rbx) + 0x50 - 0x10 == 88 == 0x58.

Did I get it right?

File: changenotes.txt (hunk 1 above)
Comment (thestr4ng3r) [reply changenotes.txt]: Yes, just for my personal notes while working.

File: librz/include/rz_analysis.h (hunk 3 above)
Comment (thestr4ng3r) [reply librz/include/rz_analysis.h]: Exactly.

**Comment by XVilka** [comment]
```
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_rebase var rebase
ANSICON=1 RZ_NOPLUGINS=1 C:\projects\rizin\rizin-vs2022_64-v0.5.0\bin\rizin.exe -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -d -Qc 'aa
afv @ entry0
doc
afv @ entry0
' bins/pe/hello_world/hello_world_od64bit.exe
-- stdout
--- expected
+++ actual
@@ -3,4 +3,4 @@
 var int64_t var_10h @ stack + 0x10
 var int64_t var_18h @ stack - 0x18
 var int64_t var_8h @ stack + 0x8
-var int64_t var_10h @ stack + 0x10EOF
+var int64_t var_10h @ stack + 0x10
-- stderr
Spawned new process with pid 1196, tid = 2000
[x] Analyze all flags starting with sym. and entry0 (aa)
```

File: librz/analysis/dwarf_process.c [hunk 4]
```diff
@@ -1700,16 +1713,22 @@ RZ_API void rz_analysis_dwarf_integrate_functions(RzAnalysis *analysis, RzFlag *
...
 				if (!i) {
 					goto loop_end;
 				}
-				rz_analysis_function_set_var(fcn, i->index, *kind, ttype, 4, false, var_name);
-			} else if (fcn) { /* kind == 'b' */
-				rz_analysis_function_set_var(fcn, offset - fcn->bp_off, *kind, ttype, 4, false, var_name);
+				RzAnalysisVarStorage stor;
+				rz_analysis_var_storage_init_reg(&stor, extra);
+				rz_analysis_function_set_var(fcn, &stor, ttype, 4, var_name);
+			} else if (fcn) { /* kind == 'b' || kind == 's' || kind == 'c' */
```
Comment (XVilka) [inline librz/analysis/dwarf_process.c]: Maybe better to convert this into a textual comment instead?

File: librz/analysis/dwarf_process.c [hunk 5]
```diff
@@ -1700,16 +1713,22 @@ RZ_API void rz_analysis_dwarf_integrate_functions(RzAnalysis *analysis, RzFlag *
...
+				RzAnalysisVarStorage stor;
+				RzStackAddr addr = offset;
+				if (*kind == 'b') {
```
Comment (XVilka) [inline librz/analysis/dwarf_process.c:1727]: What about using an enum here?

File: librz/analysis/var.c [hunk 6]
```diff
@@ -8,10 +9,30 @@
...
+	case RZ_ANALYSIS_VAR_STORAGE_STACK:
+		return b->type == RZ_ANALYSIS_VAR_STORAGE_STACK && a->stack_off == b->stack_off;
+	default:
+		rz_warn_if_reached();
+		return false;
+	}
+}
+
+/**
+ * Ensure that the register name in \p store comes from the const pool
```
Comment (XVilka) [inline librz/analysis/var.c]: Mismatch between actual parameter name and here

File: librz/analysis/var.c [hunk 7]
```diff
@@ -91,49 +87,31 @@ RZ_API void rz_analysis_var_resolve_overlaps(RzAnalysisVar *var) {
...
+		st64 otheroff = other->storage.stack_off;
+		if (strcmp(var->name, other->name) && otheroff > varoff && otheroff < varoff + varsize) {
 			rz_analysis_var_delete(other);
 		}
 	}
 	rz_pvector_free(cloned_vars);
 }
 
-RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_set_var(RzAnalysisFunction *fcn, int delta, char kind, RZ_BORROW RZ_NULLABLE const RzType *type, int size, bool isarg, RZ_NONNULL const char *name) {
+RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_set_var(RzAnalysisFunction *fcn, RZ_NONNULL RzAnalysisVarStorage *stor, RZ_BORROW RZ_NULLABLE const RzType *type, int size, RZ_NONNULL const char *name) {
```
Comment (XVilka) [inline librz/analysis/var.c:118]: Please add Doxygen here

File: librz/analysis/var.c [hunk 8]
```diff
@@ -201,12 +175,12 @@ RZ_API void rz_analysis_var_delete(RzAnalysisVar *var) {
 	}
 }
 
-RZ_API void rz_analysis_function_delete_vars_by_kind(RzAnalysisFunction *fcn, RzAnalysisVarKind kind) {
+RZ_API void rz_analysis_function_delete_vars_by_storage_type(RzAnalysisFunction *fcn, RzAnalysisVarStorageType stor) {
```
Comment (XVilka) [inline librz/analysis/var.c:192]: Doxygen.

File: librz/analysis/var.c [hunk 9]
```diff
@@ -254,89 +228,179 @@ RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_var_byname(RzAnalysisFu
 	return NULL;
 }
 
-RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_var(RzAnalysisFunction *fcn, char kind, int delta) {
+RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_var_at(RzAnalysisFunction *fcn, RZ_NONNULL RzAnalysisVarStorage *stor) {
```
Comment (XVilka) [inline librz/analysis/var.c:248]: Doxygen

File: librz/analysis/var.c [hunk 10]
```diff
@@ -254,89 +228,179 @@ RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_var_byname(RzAnalysisFu
...
 		RzAnalysisVar *var = *it;
-		if (var->kind == kind && var->delta == delta) {
+		if (storage_equals(&var->storage, stor)) {
 			return var;
 		}
 	}
 	return NULL;
 }
 
+RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_stack_var_at(RzAnalysisFunction *fcn, RzStackAddr stack_off) {
```
Comment (XVilka) [inline librz/analysis/var.c:263]: Doxygen

Comment (XVilka) [reply librz/analysis/var.c:263]: Also `rz_return_if_fail(fcn)`.

File: librz/analysis/var.c [hunk 11]
```diff
@@ -254,89 +228,179 @@ RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_var_byname(RzAnalysisFu
...
+			continue;
+		}
+		if (!best || var->storage.stack_off > best->storage.stack_off) {
+			best = var;
+		}
+	}
+	return best;
+}
+
+RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_reg_var_at(RzAnalysisFunction *fcn, const char *reg) {
```
Comment (XVilka) [inline librz/analysis/var.c]: `RZ_NONNULL` for `reg`? And `rz_return_if_fail(fcn & reg)`

File: librz/analysis/var.c [hunk 12]
```diff
@@ -1206,45 +1180,43 @@ RZ_API RZ_OWN RzList /*<RzAnalysisVar *>*/ *rz_analysis_var_list(RZ_NONNULL RzAn
...
+	}
+	return a->storage.stack_off - b->storage.stack_off;
 }
 
 static int regvar_comparator(const RzAnalysisVar *a, const RzAnalysisVar *b) {
-	// avoid NULL dereference
 	return (a && b) ? (a->argnum > b->argnum) - (a->argnum < b->argnum) : 0;
 }
 
 RZ_API void rz_analysis_fcn_vars_cache_init(RzAnalysis *analysis, RzAnalysisFcnVarsCache *cache, RzAnalysisFunction *fcn) {
```
Comment (XVilka) [inline librz/analysis/var.c:1230]: Doxygen

File: librz/core/canalysis.c [hunk 13]
```diff
@@ -5100,55 +5079,36 @@ RZ_IPI char *rz_core_analysis_var_display(RzCore *core, RzAnalysisVar *var, bool
...
-		const ut32 udelta = RZ_ABS(real_delta);
-		const char sign = real_delta >= 0 ? '+' : '-';
+		rz_strbuf_appendf(sb, "%s %s = ", rz_analysis_var_is_arg(var) ? "arg" : "var", var->name);
+	}
+	switch (var->storage.type) {
+	case RZ_ANALYSIS_VAR_STORAGE_REG: {
 		char *r;
 		if (usePxr) {
-			r = rz_core_cmd_strf(core, "pxr $w @ %s%c0x%x", analysis->reg->name[RZ_REG_NAME_BP], sign, udelta);
+			r = rz_core_cmd_strf(core, "pxr $w @r:%s", var->storage.reg);
```
Comment (XVilka) [inline librz/core/canalysis.c:5094]: Please add here `// TODO: Convert to the API`

File: librz/core/canalysis.c [hunk 14]
```diff
@@ -5100,55 +5079,36 @@ RZ_IPI char *rz_core_analysis_var_display(RzCore *core, RzAnalysisVar *var, bool
...
 		} else {
-			r = rz_core_cmd_strf(core, "pf %s @ %s%c0x%x", fmt, analysis->reg->name[RZ_REG_NAME_BP], sign, udelta);
+			r = rz_core_cmd_strf(core, "pf r (%s)", var->storage.reg);
```
Comment (XVilka) [inline librz/core/canalysis.c:5097]: And here

File: librz/core/canalysis.c [hunk 15]
```diff
@@ -5100,55 +5079,36 @@ RZ_IPI char *rz_core_analysis_var_display(RzCore *core, RzAnalysisVar *var, bool
...
-	case RZ_ANALYSIS_VAR_KIND_SPV: {
-		ut32 udelta = RZ_ABS(var->delta + var->fcn->maxstack);
+		break;
+	}
+	case RZ_ANALYSIS_VAR_STORAGE_STACK: {
+		ut64 addr = rz_analysis_var_addr(var);
 		char *r;
 		if (usePxr) {
-			r = rz_core_cmd_strf(core, "pxr $w @ %s+0x%x", analysis->reg->name[RZ_REG_NAME_SP], udelta);
+			r = rz_core_cmd_strf(core, "pxr $w @ 0x%" PFMT64x, addr);
```
Comment (XVilka) [inline librz/core/canalysis.c:5108]: And here

File: librz/core/canalysis.c [hunk 16]
```diff
@@ -5100,55 +5079,36 @@ RZ_IPI char *rz_core_analysis_var_display(RzCore *core, RzAnalysisVar *var, bool
...
 		} else {
-			r = rz_core_cmd_strf(core, "pf %s @ %s+0x%x", fmt, analysis->reg->name[RZ_REG_NAME_SP], udelta);
+			r = rz_core_cmd_strf(core, "pf %s @ 0x%" PFMT64x, fmt, addr);
```
Comment (XVilka) [inline librz/core/canalysis.c:5111]: And here

File: librz/core/cmd/cmd_analysis.c [hunk 17]
```diff
@@ -2627,29 +2590,29 @@ RZ_IPI RzCmdStatus rz_analysis_function_vars_handler(RzCore *core, int argc, con
...
 	case RZ_OUTPUT_MODE_STANDARD:
-		core_analysis_var_list_show(core->analysis, fcn, RZ_ANALYSIS_VAR_KIND_SPV, state);
-		core_analysis_var_list_show(core->analysis, fcn, RZ_ANALYSIS_VAR_KIND_BPV, state);
-		core_analysis_var_list_show(core->analysis, fcn, RZ_ANALYSIS_VAR_KIND_REG, state);
+		core_analysis_var_list_show(core->analysis, fcn, RZ_ANALYSIS_VAR_STORAGE_STACK, state);
+		core_analysis_var_list_show(core->analysis, fcn, RZ_ANALYSIS_VAR_STORAGE_REG, state);
 		break;
 	case RZ_OUTPUT_MODE_RIZIN: {
 		bp = rz_reg_get_name(core->analysis->reg, RZ_REG_NAME_BP);
 		rz_cons_printf("f-fcnvar*\n");
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c:2598]: Please fix this too while you are on it. `"f- fcnvar*"`

File: librz/core/cmd/cmd_analysis.c [hunk 18]
```diff
@@ -2918,83 +2872,96 @@ static RzCmdStatus analysis_function_vars_del(RzCore *core, RzAnalysisVarKind ki
...
 		return RZ_CMD_STATUS_ERROR;
 	}
 
 	RzAnalysisOp *op = rz_core_analysis_op(core, addr, 0);
 	const char *ireg = op ? op->ireg : NULL;
-	if (kind == RZ_ANALYSIS_VAR_KIND_SPV) {
-		delta -= fcn->maxstack;
+	st64 addend = 0;
+	if (stor->type == RZ_ANALYSIS_VAR_STORAGE_STACK) {
+		// TODO: this is wrong. What we need is not the address on the stack, but
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c:2907]: Probably `FIXME` not `TODO` then

File: librz/parse/p/parse_arm_pseudo.c [hunk 19]
```diff
@@ -205,71 +205,117 @@ static bool parse(RzParse *p, const char *assembly, RzStrBuf *sb) {
...
+	int group_idx_sign = -1;
+	int group_idx_addend = 2;
+	bool brackets = true;
+	if (p->pseudo) {
+		// match e.g. "fp - 0x42"
+		// capturing "fp", "-", "0x42"
+		re_str = "\\[([a-z][0-9a-z][0-9]?)\\s*(\\+|-)\\s*(-?(0x)?[0-9a-f]+)\\]";
+		group_idx_sign = 2;
+		group_idx_addend = 3;
+	} else if (op && op->type == RZ_ANALYSIS_OP_TYPE_ADD && op->dst && op->dst->reg && op->dst->reg->name && rz_str_casecmp(op->dst->reg->name, "fp") && rz_str_casecmp(op->dst->reg->name, "sp")) {
```
Comment (XVilka) [inline librz/parse/p/parse_arm_pseudo.c]: Maybe split this or extract?

File: librz/parse/p/parse_mips_pseudo.c [hunk 20]
```diff
@@ -141,110 +141,103 @@ static bool parse(RzParse *parse, const char *assembly, RzStrBuf *sb) {
...
+	if (!p->pseudo) {
+		// match e.g. -0x18(%rbp)
+		// capturing "-0x18", "0x", "rbp"
+		re_str = "(-?(0x)?[0-9a-f]+)\\(([a-z][0-9a-z]))";
+		group_idx_reg = 3;
+		group_idx_sign = -1;
+		group_idx_addend = 1;
+	} else {
+		// match e.g. rbp - 0x42
+		// capturing "rbp", "-", "0x42"
```
Comment (XVilka) [inline librz/parse/p/parse_mips_pseudo.c]: What `rbp` does in MIPS?

File: librz/parse/p/parse_mips_pseudo.c [hunk 21]
```diff
@@ -141,110 +141,103 @@ static bool parse(RzParse *parse, const char *assembly, RzStrBuf *sb) {
...
-			tmpf = "0x%x(%s)";
+	}
+
+	char *varstr = p->var_expr_for_reg_access(f, addr, reg_str, reg_addend);
+	if (!varstr) {
+		free(reg_str);
+		return tstr;
+	}
+
+	// replace!
```
Comment (XVilka) [inline librz/parse/p/parse_mips_pseudo.c]: Replace with what?

File: librz/util/json_parser.c [hunk 22]
```diff
@@ -540,22 +540,30 @@ static void json_pj_recurse(const RzJson *json, PJ *pj, bool with_key) {
 	}
 }
 
+/**
+ * \brief Print the contents of \p json into \p pj
+ * \param json the data to read from
+ * \param pj the PJ to print into
+ * \param with_key whether to include the root key of \p json, i.e. `"key": <val>`, vs. `<val>`
+ */
+RZ_API void rz_json_to_pj(const RzJson *json, PJ *pj, bool with_key) {
```
Comment (XVilka) [inline librz/util/json_parser.c]: Please add `RZ_NONNULL`

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/analysis/dwarf_process.c (hunk 5 above)
Comment (thestr4ng3r) [reply librz/analysis/dwarf_process.c:1727]: Could be done, but this comes from sdb and should rather be refactored completely to not use sdb at all.

File: librz/analysis/var.c (hunk 12 above)
Comment (thestr4ng3r) [reply librz/analysis/var.c:1202]: I will deprecate it. This does not make much sense anymore.

File: librz/core/cmd/cmd_analysis.c (hunk 18 above)
Comment (thestr4ng3r) [reply librz/core/cmd/cmd_analysis.c:2907]: Did we ever specify any difference between `FIXME` and `TODO`? I personally always write `TODO`.

Comment (XVilka) [reply librz/core/cmd/cmd_analysis.c:2907]: Not really. Not that important anyway.

**Review by XVilka — APPROVED** [review]

**Review by wargio — APPROVED** [review]

**Pull request merged**